### PR TITLE
Defining a default target to support cross-compilation targets 

### PR DESCRIPTION
**Pull request by FuriouZz**
Related to #4163

### CODE REVIEW

File: editors/code/package.json [hunk 1]
```diff
@@ -237,6 +237,14 @@
                     "default": false,
                     "markdownDescription": "Run `cargo check` on startup to get the correct value for package OUT_DIRs"
                 },
+                "rust-analyzer.cargo.rustc.defaultTarget": {
```
Comment (bjorn3) [inline editors/code/package.json:240]:
```suggestion
                "rust-analyzer.rustc.defaultTarget": {
```

or

```suggestion
                "rust-analyzer.defaultTarget": {
```

Comment (FuriouZz) [reply editors/code/package.json:240]: This is because I put the option into the `CargoConfig`. Maybe `rust-analyzer.cargo.defaultTarget` ?

**Comment by matklad** [comment]
LGTM, but let's call this `filterPlatform` rather than `defaultTarget`, to not deviate from cargo's command line

**Comment by FuriouZz** [comment]
@matklad I do the change.
We do not need to do any change to the rustc command here ? https://github.com/rust-analyzer/rust-analyzer/blob/fc57358efda7c028cbe8a438446cce5f540f48ca/crates/ra_project_model/src/lib.rs#L561 ?

**Comment by matklad** [comment]
Actually, i think we do need to change that as well, good catc!

**Comment by matklad** [comment]
Hm, though then `filterTarget` kind of becomes a bad name. Let's just use target?

**Comment by FuriouZz** [comment]
@matklad The only problem is that I don't know how to access cargo options to this point =/.

**Comment by matklad** [comment]
I think adding `fn get_rustc_cfg_options(target: Option<String>)` and then fixing call sites should work. `CargoConfig` would be a part of bigger `Config`.

**Comment by FuriouZz** [comment]
> Hm, though then `filterTarget` kind of becomes a bad name. Let's just use target?

By default, rust-analyzer check all targets right? `filterTarget` can be good. `--filter-platform` can accept multiple targets?

**Comment by matklad** [comment]
>By default, rust-analyzer check all targets right?

Yes, though this is handled by completely different cargo instance: 

https://github.com/rust-analyzer/rust-analyzer/blob/aef95ff688391065082e2e67040cba2201922005/crates/ra_flycheck/src/lib.rs#L26-L30

I think let's use `cargo.target` for now, and then maybe reconsider later :D

File: crates/ra_project_model/src/lib.rs [hunk 2]
```diff
@@ -558,10 +558,18 @@ pub fn get_rustc_cfg_options() -> CfgOptions {
...
+            Command::new("rustc")
+                .args(&["--print", "cfg", "-O", "--target", target.as_str()])
+                .output()
+                .context("Failed to get output from rustc --print cfg -O")?
+        } else {
+            Command::new("rustc")
+                .args(&["--print", "cfg", "-O"])
+                .output()
+                .context("Failed to get output from rustc --print cfg -O")?
+        };
```
Comment (matklad) [inline crates/ra_project_model/src/lib.rs:571]: Let's remove code duplicaiton here:

```
let mut cmd = Command::new("rustc");
cmd.args(&[...])
if let Some(target) = target { cmd.args(&["--target", target]) }
cmd.oupput.context()

```

Comment (FuriouZz) [reply crates/ra_project_model/src/lib.rs:571]: oups, you're right

**Comment by matklad** [comment]
bors r+

Thanks!

**Comment by matklad** [comment]
cc @jonas-schievink that's something which might be interesting for embedded stuff

**Pull request merged**